mava/wrappers/gigastep.py

```diff
@@ -0,0 +1,301 @@
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import TYPE_CHECKING, Dict, Tuple, Union
+
+import jax
+import jax.numpy as jnp
+from chex import Array, PRNGKey
+from gigastep.gigastep_env import GigastepEnv
+from jumanji import specs
+from jumanji.types import StepType, TimeStep, restart
+from jumanji.wrappers import Wrapper
+
+from mava.types import Observation, ObservationGlobalState, State
+
+if TYPE_CHECKING:  # https://github.com/python/mypy/issues/6239
+    from dataclasses import dataclass
+else:
+    from flax.struct import dataclass
+
+
+@dataclass
+class GigastepState:
+    """Wrapper around a Gigastep state to provide necessary attributes for Jumanji environments."""
+
+    state: State
+    key: PRNGKey
+    step: int
+    adversary_action: Array
+
+
+class GigastepWrapper(Wrapper):
+    """Wraps a Gigastep environment so that its API is compatible with Jumanji environments."""
+
+    def __init__(
+        self,
+        env: GigastepEnv,
+        has_global_state: bool = False,
+    ):
+        """
+        Args:
+            env: The Gigastep environment to be wrapped.
+            time_limit (int): The maximum duration of each episode, in seconds. Defaults to 500.
+            has_global_state (bool): Whether the environment has a global state. Defaults to False.
+        """
+
+        super().__init__(env)
+        assert (
+            env.discrete_actions
+        ), "Only discrete action spaces are currently supported for Gigastep environments"
+        assert (
+            env._obs_type == "vector"
+        ), "Only Vector observations are currently supported for Gigastep environments"
+
+        self._env: GigastepEnv
+        self._timelimit = self._env.max_episode_length
+        self.num_agents = self._env.n_agents_team1
+        self.num_actions = self._env.n_actions
+
+        self.has_global_state = has_global_state
+
+    def reset(self, key: PRNGKey) -> Tuple[GigastepState, TimeStep]:
+        """
+        Reset the Gigastep environment.
+
+        Args:
+            key (PRNGKey): The PRNGKey.
+
+        Returns:
+            GigastepState : the state of the environment.
+            TimeStep : the first time step.
+
+        """
+
+        key, reset_key, adversary_key = jax.random.split(key, 3)
+        obs, state = self._env.reset(reset_key)
+
+        obs_team1, state_team1, obs_team2, state_team2 = self._split_obs_and_state(obs, state)
+
+        adversary_actions = self.adversary_policy(obs_team2, state_team2, adversary_key)
+
+        state = GigastepState(state, key, 0, adversary_actions)
+
+        obs = self._create_observation(obs_team1, adversary_actions, obs, state)
```

Comment:
yes, since the old observation is not passed to the `step()` function the adversary needs to decide it's actions right when the observation is created. that decision is then stored in the state and only executed when `step()` is called.

Reply:
When the trained team decide based on the immediate taken action of the adversary seems like a new other approach, not a decentralized execution but something like sequential decision taken (not exactly sequential). 
So, my previous suggestion was to choose whether or not included the `last actions` of all agents or only adversary not the `immediate actions` (and please check the smacv2 implementation).

Reply:
In this case, I would suggest initializing this in the reset() with infinite values(outer of range values) and then in the step() use the immediate values for the next obs creation.